From: David Teigland <teigland@redhat.com>
Date: Wed, 3 Sep 2008 09:02:21 -0500
Subject: [dlm] fix address compare
Message-id: 20080903140221.GD22775@redhat.com
O-Subject: [RHEL5.3 PATCH] dlm: fix address compare
Bugzilla: 459585
RH-Acked-by: Christine Caulfield <ccaulfie@redhat.com>

bz 459585 dlm_recoverd in D state when using IPv6 to comunicate

brew build including this patch
http://brewweb.devel.redhat.com/brew/taskinfo?taskID=1451582

upstream commit (in linux-next for 2.6.28):

Author: David Teigland <teigland@redhat.com>
Date:   Thu Aug 28 11:36:19 2008 -0500

    dlm: fix address compare

    Compare only the addr and port fields of sockaddr structures.
    Fixes a problem with ipv6 where sin6_scope_id does not match.

    Signed-off-by: David Teigland <teigland@redhat.com>

diff --git a/fs/dlm/config.c b/fs/dlm/config.c
index 5069b2c..a7b48c5 100644
--- a/fs/dlm/config.c
+++ b/fs/dlm/config.c
@@ -14,6 +14,9 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/configfs.h>
+#include <linux/in.h>
+#include <linux/in6.h>
+#include <net/ipv6.h>
 #include <net/sock.h>
 
 #include "config.h"
@@ -771,6 +774,33 @@ static void put_space(struct space *sp)
 	config_item_put(&sp->group.cg_item);
 }
 
+static int addr_compare(struct sockaddr_storage *x, struct sockaddr_storage *y)
+{
+	switch (x->ss_family) {
+	case AF_INET: {
+		struct sockaddr_in *sinx = (struct sockaddr_in *)x;
+		struct sockaddr_in *siny = (struct sockaddr_in *)y;
+		if (sinx->sin_addr.s_addr != siny->sin_addr.s_addr)
+			return 0;
+		if (sinx->sin_port != siny->sin_port)
+			return 0;
+		break;
+	}
+	case AF_INET6: {
+		struct sockaddr_in6 *sinx = (struct sockaddr_in6 *)x;
+		struct sockaddr_in6 *siny = (struct sockaddr_in6 *)y;
+		if (!ipv6_addr_equal(&sinx->sin6_addr, &siny->sin6_addr))
+			return 0;
+		if (sinx->sin6_port != siny->sin6_port)
+			return 0;
+		break;
+	}
+	default:
+		return 0;
+	}
+	return 1;
+}
+
 static struct comm *get_comm(int nodeid, struct sockaddr_storage *addr)
 {
 	struct config_item *i;
@@ -792,8 +822,7 @@ static struct comm *get_comm(int nodeid, struct sockaddr_storage *addr)
 			config_item_get(i);
 			break;
 		} else {
-			if (!cm->addr_count ||
-			    memcmp(cm->addr[0], addr, sizeof(*addr)))
+			if (!cm->addr_count || !addr_compare(cm->addr[0], addr))
 				continue;
 			found = 1;
 			config_item_get(i);
